engine/src/main/java/com/datatorrent/stram/StreamingContainerManager.java

```diff
@@ -3208,10 +3217,63 @@ public Object call() throws Exception
     return null;
   }
 
+  private class ChangeDAGRunnable implements Callable<Object>
+  {
+    private final transient DAGChangeSetImpl request;
+
+    public ChangeDAGRunnable(DAGChangeSetImpl changes)
+    {
+      request = changes;
+    }
+
+    @Override
+    public Object call() throws Exception
+    {
+      if (request == null) {
+        return null;
+      }
+
+      try {
+        LOG.info("starting plan change ");
+        LogicalPlan lp = cloneLogicalPlan();
+        PlanModifier pm = new PlanModifier(lp);
+        pm.applyDagChangeSet(request);
+        lp.validate();
+
+        LOG.info("validated logical plan, now making actual change");
+        pm = new PlanModifier(plan);
+        pm.applyDagChangeSet(request);
+        LOG.info("plan change done");
+        return null;
+      } catch (Exception ex) {
+        // TODO send exception back to stat listener for taking action.
+        LOG.error("Exception occured {}", ex);
+        throw ex;
+      }
+    }
+  }
+
   @VisibleForTesting
   protected Object getLogicalCounter(String operatorName)
   {
     return latestLogicalCounters.get(operatorName);
   }
 
+  public FutureTask<Object> logicalPlanModification(DAGChangeSetImpl dag) throws Exception
+  {
+    FutureTask<Object> future = new FutureTask<>(new ChangeDAGRunnable(dag));
+    dispatch(future);
+    return future;
+  }
+
+  @Override
+  public void addDagChangeRequests(DAG dag)
+  {
+    try {
+      logicalPlanModification((DAGChangeSetImpl)dag);
+    } catch (Exception ex) {
+      // pass
```

Comment:
I am working on way to provide result of dag modifications back to stats listener so that they could take decision based on it instead of ignoring the exception.